### PR TITLE
refactor: add pipeline transform duplex streams

### PR DESCRIPTION
**Pull request by ricellis**
## Checklist

- [x] Added tests for code changes _or_ test/build only changes
- [ ] Updated the change log file (`CHANGES.md`) _or_ test/build only changes - merging to `modernization`, changes will be updated at end of modernization work
- [x] Completed the PR template below:

## Description

Add pipeline transform duplex streams.

Fixes i250 i253 i254 i255

## Approach

* Add `transforms.js` with a series of transforming duplex streams
    * Add a BatchingStream for batching elements into arrays
    * Add a SplittingStream for flat-mapping a batch back to elements
    * Add a MappingStream for mapping stream elements to something else

## Schema & API Changes

- "No change"

## Security and Privacy

- "No change"

## Testing

- Added new unit tests for each transform stream type

## Monitoring and Logging

- "Added new logging group `couchbackup:transform` with subgroup `batch` for debug of the batching implementation.

### CODE REVIEW

File: test/transforms.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    it('multiple batches, same size', async function() {
+      return testSplitting(15, 3);
+    });
+
+    it('multiple batches, different size', async function() {
+      return testSplitting(29, 8);
+    });
+
+    it('multiple batches, concurrency', async function() {
```
Comment (mojito317) [inline test/transforms.js:103]: Maybe a test case with multiple batches, different size, and concurrency would be worth it to cover edge cases.

File: includes/transforms.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    if (this.batch.push(element) === this.batchSize) {
+      this.writeBatch(callback);
+    } else {
+      callback();
+    }
+  }
+
+  _flush(callback) {
+    this.log('Flushing batch transform.');
+    this.writeBatch(callback);
```
Comment (mojito317) [inline includes/transforms.js:53]: A comment here would be recommended as to why this is called for the last batch.

**Review by mojito317 — APPROVED** [review]
Overall this looks good to me! I have 2 minor comments.

File: includes/transforms.js (hunk 2 above)
Comment (ricellis) [reply includes/transforms.js:56]: added in https://github.com/IBM/couchbackup/pull/643/commits/3890dac2721963a407173466349af3c0d348aa91

**Review by emlaver — APPROVED** [review]
